Specify and document current PM_DEVICE behavior, including PM_DEVICE=n and device_deinit()

bjarki-andreasen opened the pull request:
This PR aims to document the status quo of device power management to provide a foundation for current driver development, and future refactoring discussions, since we need to agree on the current state before being able to discuss the future state :)

The PR also adds the missing counterpart to `pm_device_driver_init()`.

bjarki-andreasen-nordic committed:
pm: device: introduce pm_device_driver_deinit()

Introduce pm_device_driver_deinit() which is required to fully
support device_deinit(). This function shall be called by drivers
when device_deinit() is called.

If PM_DEVICE is enabled, it will
simply check whether the device is either SUSPENDED or OFF, this
will prevent device_deinit() on a device which is currently in
use, and ensure the device is in the correct state if device_init()
is called later.

If PM_DEVICE is not enabled, it will invoke the SUSPEND action which
mirrors the pm_device_driver_init() behavior. Note that TURN_OFF
action is not called since the device is deinitialized after this
call.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen-nordic committed:
tests: pm: device_driver_init: extend to support device_deinit()

Extend test suite to cover pm_device_driver_deinit() resulting
from call to device_deinit().

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

JordanYates reviewed (commented):
What is the purpose of this function? It seems odd that it is running functions when `CONFIG_PM_DEVICE=n` but doesn't do anything when `CONFIG_PM_DEVICE=y`?

JordanYates commented on subsys/pm/device.c:
```diff
@@ -401,3 +401,14 @@ int pm_device_driver_init(const struct device *dev,
 	return rc;
 }
+
+int pm_device_driver_deinit(const struct device *dev,
```
This is still confusing to me. If `PM_DEVICE=n`, the function implementation actually does work of de-initialising the device. If `PM_DEVICE=y`, the function is a boolean check of whether it is ready to be de-initialised?

Same function, two very different meanings as far as I can tell.

bjarki-andreasen replied:
that is correct, the same is the case for pm_device_driver_init():
* If `PM_DEVICE=n`, call on action followed by resume action
* If `PM_DEVICE=y`, init the PM device structure and resume device
* If `PM_DEVICE_RUNTIME=y` and `PM_DEVICE=y` and device has runtime enabled, init the PM device structure but keep device suspended.

We need a common path to deinit/suspend the device if `PM_DEVICE=n`. if `PM_DEVICE=y`, then the user will already have suspended the device with pm_device_action_run() before calling device_deinit(), hence we just need to check that the user actually did this :)

bjarki-andreasen replied:
Note that `pm_device_driver_deinit()` will only be called as a result of `device_deinit()`, which forces the driver to release all resources, including what was claimed from `ACTION_TURN_ON`. The device will be left in a deinitialized state, meaning `device_is_ready()` returns `false` after `pm_device_driver_deinit()` has been called.

JordanYates replied:
> Note that pm_device_driver_deinit() will only be called as a result of device_deinit(), which forces the driver to release all resources, including what was claimed from ACTION_TURN_ON

How is `device_deinit` supposed to know which resources to release? We removed `pm_device_state_get` from drivers for a good reason.

bjarki-andreasen replied:
A device should not acquire any resources outside of the state `ACTIVE`, and it should release all resources in `SUSPENDED`. The state `OFF` is just there to inform the driver that when this device is turned on again, all hardware registers etc. have been cleared, and you need to reinit them. There is no `ON` state, its just a transition from OFF to suspended, used to initialize the hardware.

bjarki-andreasen replied:
Check out **Device Behavior for Suspend and Turn Off** section of https://github.com/zephyrproject-rtos/zephyr/issues/84049

JordanYates reviewed (changes requested):
> For now, this is a simple check ensuring the device is not in use

What is stopping a final implementation from being implemented now? The function should have defined semantics, and they should be implemented properly when the function is introduced, not modified at some point in "the future".

bjarki-andreasen commented:
> > For now, this is a simple check ensuring the device is not in use
> 
> What is stopping a final implementation from being implemented now? The function should have defined semantics, and they should be implemented properly when the function is introduced, not modified at some point in "the future".

This is all we need for now, I don't think there is a "final" implementation, its just a hook for doing whatever is required of PM :)

JordanYates commented on include/zephyr/pm/device.h:
```diff
@@ -731,6 +742,23 @@ static inline int pm_device_driver_init(const struct device *dev, pm_device_acti
+		return rc;
+	}
+
+	rc = action_cb(dev, PM_DEVICE_ACTION_TURN_OFF);
```
This doesn't seem right, the device is not literally being turned off, just the software construct.
This would lead to GPIO's being put into incorrect states if used generically, which the PM API is.

Overloading the meaning of `PM_DEVICE_ACTION_TURN_OFF` is a bad idea IMO.

bjarki-andreasen replied:
I'm okay with removing the call to `TURN_OFF`, given that all resources like GPIOs are released as part of `device_deinit()`, so I think you are right that makes little sense to act as if the driver is being turned off before being deinitialized.

ceolin replied:
Agree about it

ceolin commented on include/zephyr/pm/device.h:
```diff
@@ -635,6 +635,17 @@ bool pm_device_is_powered(const struct device *dev);
+/**
+ * @brief Deinit a device driver
+ *
+ * This function must be called at the end of a driver deinit function.
```
Shouldn't it be called at the beginning of the de-initialization ? Since if the device is active it will return -EBUSY and the de-initialization should be postponed ?

bjarki-andreasen replied:
yes, will update description :)

bjarki-andreasen replied:
done

bjarki-andreasen commented:
@JordanYates could you revisit this one?

tobiwan88 commented on subsys/pm/device.c:
```diff
@@ -401,3 +401,14 @@ int pm_device_driver_init(const struct device *dev,
+{
+	struct pm_device_base *pm = dev->pm_base;
+
+	return pm->state == PM_DEVICE_STATE_SUSPENDED ||
```
I would say suspend and off are not the same. 

Looking at this #84049  suspend could mean that the device is still consuming a resource, maybe only a power saving mode.  I know that e.g. nordic does a full deinit of the resource in suspended, but at least from the state diagrams and the description of the issue, the behaviour from suspended and off can be differently and there is no gurantee that a suspended the device is really "off" and claimed resources can be reused again

bjarki-andreasen replied:
This scenario is simply not handled by device drivers. We discussed it with https://github.com/zephyrproject-rtos/zephyr/pull/75922 in the past. Whatever combination of GPIO and pinctrl states required to prevent any leakage/backpowering needs to be handled by some higher level construct. This complexity should not be copied to every single device driver.

In this specific scenario, in suspended, GPIO would be pulled high, in OFF, GPIO would be left where it is. Not optimal, but this usecase is simply not handled yet. Usually active low pins have pull ups, and in "real" applications, GPIOs are usually just forced to appropriate state by the application before sleep (ugly, but works and we don't have anything better in zephyr ATM)

JordanYates replied:
I disagree with most of that, I wrote a chunk of text, but it doesn't really matter in the context of this PR.

> SUSPEND and OFF from a resource perspective are the same

Why is this true? I cannot see anywhere that specifies that no resources are allowed to be claimed in the `SUSPEND` state. It is trivial to construct an example where a resource *does* need to be claimed in `SUSPEND` mode (a device that is either running or idle depending on a control GPIO), so simply claiming that `SUSPEND` is not allowed to claim resources cannot be used as a justification.

bjarki-andreasen replied:
Did you _"Check out Device Behavior for Suspend and Turn Off section of https://github.com/zephyrproject-rtos/zephyr/issues/84049"_ ? I'm not "just claiming" anything, at worst I'm misunderstanding something. To look to existing documentation:
https://github.com/zephyrproject-rtos/zephyr/blob/11c543e24cddeab1b393fbbec88c5517cb4d4456/include/zephyr/pm/device.h#L71-L77 and
https://github.com/zephyrproject-rtos/zephyr/blob/11c543e24cddeab1b393fbbec88c5517cb4d4456/include/zephyr/pm/device.h#L80-L86 and
https://github.com/zephyrproject-rtos/zephyr/blob/11c543e24cddeab1b393fbbec88c5517cb4d4456/include/zephyr/pm/device.h#L95-L100 OFF == context was lost because power was removed, that's **the** difference between OFF and SUSPENDED. There will be cases where some resources may be kept "requested" in OFF state, and drivers are allowed to do that, otherwise why even invoke TURN_OFF action. In these cases, device_deinit() would probably be the place to release those.

JordanYates reviewed (changes requested):
It's still not clear to me how this helps to reliably implement the requirements of the `device_deinit` API.
https://github.com/zephyrproject-rtos/zephyr/blob/af46f628a0c5b21fca3ac4a78640059f3cb19442/include/zephyr/device.h#L880-L885

It's on the caller to determine whether the device can be de-initialized, not the implementation.

Why is `STATE_SUSPENDED` any more likely to be the "reset state" of a device than `STATE_ACTIVE`? In many cases the suspend state is explicitly not the reset state of a device.

JordanYates commented:
As a concrete request, please document what the function is actually supposed to be doing beyond just `@brief Deinit a device driver`. It is easier to check that an intended behavior makes sense and that an implementation matches the intended behavior than to try and reason about how a particular implementation is supposed to fit into the larger `device_deinit` API based only on code.

bjarki-andreasen-nordic committed:
pm: device: extend pm_device_driver_init() description

Rewrite the docstring for pm_device_driver_init() to align with
documentation, especially covering the case where CONFIG_PM_DEVICE=n

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

JordanYates commented on include/zephyr/pm/device.h:
```diff
@@ -628,13 +628,33 @@ bool pm_device_is_powered(const struct device *dev);
+/**
+ * @brief Deinit a device driver
+ *
+ * @details Ensures device is either SUSPENDED or OFF, and will revert the actions of
```
This says it is reverting the actions of `pm_device_driver_init`, but it is only reverting half of them per your own additions.

JordanYates replied:
IMO, the docs as they are do not address the spirit of the request.

> As a concrete request, please document what the function is actually supposed to be doing beyond just @brief Deinit a device driver. It is easier to check that an intended behavior makes sense and that an implementation matches the intended behavior than to try and reason about how a particular implementation is supposed to fit into the larger device_deinit API based only on code.

Just listing the functions that it calls internally does not help anyone understand the role of the function. Asking me for suggestions is pointless because I am still perplexed by the role of a function that reverts different pieces of code depending on kconfig options, so how am I supposed to know how to document it?

Why is it correct for the driver to be in the equivalent of `PM_DEVICE_STATE_OFF` OR `PM_DEVICE_STATE_SUSPENDED` when this function returns. I would love for the docs to be something along the lines of
> Releases all the resources claimed by the PM transitions towards `PM_DEVICE_STATE_ACTIVE` by running the transitions back to `PM_DEVICE_STATE_OFF`, reversing `pm_device_state_init`.

Which is clear about the goal and expected end state of the function, but it obviously is not what the function currently does.

bjarki-andreasen-nordic committed:
doc: services: pm: update pm device init/deinit/states

- Specify "Device" Power Management Support in sections covering
  CONFIG_PM_DEVICE=y or n
- Extend the enum pm_device_state docstrings to define and detail
  expectations of devices in each enum pm_device_state state.
- Extend pm device driver code example with more concise comments
  and include pm_device_driver_init, pm_device_driver_deinit, and
  DEVICE_DT_INST_DEINIT_DEFINE() (device deinit feature)
- Describe the device driver PM states used if PM_DEVICE is not
  enabled.
- Note which states a device driver is initialized from and can
  be deinitialized in and why.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,39 @@ a particular device. It is important to emphasize that, although the
+
+- ``OFF``: Device hardware is not powered. Once this state is exited, the device
+  is considered to be in its reset state. Every device is assumed to be in this
+  state once a device driver is initialized, as we assume it has been
```
Reworded it to be more clear :)

JordanYates commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,42 @@ a particular device. It is important to emphasize that, although the
+  system. The device should be disabled in this state. No device driver API
+  calls will occur in this state. The device will be powered off if its power
+  domain is no longer needed by the system. Note that this state is **NOT** a
+  "low-power"/"partially operable" state, those are configured using device
```
I think I understand what you are getting at here , but it should be clarified that while this isn't a low power "operation" state, the device **should** be configured into its lowest power internal state when entering `SUSPENDED`.

ceolin replied:
That makes sense to me

JordanYates commented on doc/services/pm/device.rst:
```diff
@@ -202,58 +238,126 @@ power management implementation.
 
-Here is some example code showing how to implement device power management
-support in a device driver.
+The following example code shows how to implement device power management
```
I think it would be extremely useful for this to actually include some example resources that are commonly controlled in drivers (probably without error handling for brevity). As it is, the example shows the form of the callback, but doesn't really show what an implementation actually does.

I would propose two hypothetical GPIOs:
 *  One that is configured to be used as an interrupt in `ACTIVE`
 * One that controls the low power state of the device (Device is in a low power sleep mode when `GPIO_OUTPUT_ACTIVE`).

bjarki-andreasen replied:
I extended the driver example to be, detailed, maybe a bit overkill even.

JordanYates replied:
That is great, better than I hoped for.

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,42 @@ a particular device. It is important to emphasize that, although the
+  "low-power"/"partially operable" state, those are configured using device
+  driver specific APIs and apply only while the device is ``ACTIVE``.
+
+  A device driver may be deinitialized in this state as the device will be
```
Extended documentation to make this clearer

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -202,58 +241,290 @@ power management implementation.
+           (void)gpio_remove_callback(config->int_pin.port, &data->int_pin_callback);
+
+           /* Disable the device. In this case, we use the enable pin */
+           (void)gpio_pin_set_dt(&config->enable_pin, 0);
```
Right, I actually had a comment specifically mentioning that GPIOs are "assumed to retain pin configurations even as we release them", which I later decided against. I have extended the code example quite a bit instead with explicit assumptions regarding GPIO pins and other "persistant configurations"

bjarki-andreasen replied:
Yes, compare it to the device and GPIOs states after soc reset. We know if the hardware is powered when initializing the device driver, but we have no idea if the board was power cycled or just reset. GPIOs will be in whatever their reset state is, and the device will be in an unknown state. The states OFF and SUSPENDED are analogous to the state the optimal states the device will be in when initializing it, though an soc reset could leave the device ACTIVE as well if there was a hardfault :)

JordanYates commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,45 @@ a particular device. It is important to emphasize that, although the
+  deinitialized device driver. If the device was left in the ``SUSPENDED`` state,
+  the :c:enum:`pm_device_action` ``TURN_ON`` action will be invoked whether or
+  not the device hardware lost power.
+
```
The above all looks technically correct, but is pretty dense in terms of what a user needs to actually do.
Maybe add something like the following?
```
As a result, `PM_DEVICE_ACTION_TURN_ON` should be able to transition the device from any internal state into `SUSPENDED`, since no guarantees can be made across resets. This is typically achieved through toggling a reset pin or triggering a software reset through a register write before performing any additional configuration needed to meet the requirements of ``SUSPENDED``.
```

bjarki-andreasen replied:
Added this unaltered aside from .rst formatting, removed preceeding duplicate information, appended

```
For devices where this is not possible, the device driver
must presume the device is in either the ``OFF`` or ``SUSPENDED`` state at
time of initialization, as these are the states within which device drivers
may be deinitialized.
```
since this is the case for some nordic hardware, we don't have a way of resetting UARTs outside of system resets, so bootloaders need to do the deinit properly, and app needs to presume bootloader nicely reset the hardware :)

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,45 @@ a particular device. It is important to emphasize that, although the
+
+  A device driver may be deinitialized in this state. Once the device driver
+  has been deinitialized, we implicitly move to the ``OFF`` state as the device
+  hardware may loose power, with no device driver to respond to the
```
done

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -202,58 +241,308 @@ power management implementation.
+            * once the power domain is suspended, so we configure it as an
+            * input.
+            */
+           (void)gpio_pin_configure_dt(&config->enable_pin, GPIO_INPUT);
```
updated to
```
           /*
            * We prepare the device for being powered off. In this case, we
            * have an active low enable pin, which could back power the device
            * once the power domain is suspended, so we configure it as
            * disconnected if supported, input otherwise.
            */
           if (gpio_pin_configure_dt(&config->enable_pin, GPIO_DISCONNECTED)) {
                   (void)gpio_pin_configure_dt(&config->enable_pin, GPIO_INPUT);
           }
```

bjarki-andreasen commented:
@decsny I added an entry to cover the case you mentioned here https://github.com/zephyrproject-rtos/zephyr/pull/90946#issuecomment-2970872930

specifically in the example driver's resume function, I added:
```
           /*
            * Release devices currenty not needed by device. In this case, we
            * are releasing the bus and the enable pin.
            *
            * The device driver would keep the bus ACTIVE while the device is
            * ACTIVE in cases of high throughput or unsolicitet data on the
            * bus. To avoid inefficient RESUME/SUSPEND cycles of the bus
            * for every transaction, and allowing reception of unsolicitet
            * data on serial busses like UART.
            */
```

bjarki-andreasen commented:
@JordanYates could you take another look?

bjarki-andreasen commented:
@decsny could you take a look?

ceolin reviewed (dismissed):
Thanks ! It looks great!

nashif commented on include/zephyr/pm/device.h:
```diff
@@ -638,6 +638,23 @@ bool pm_device_is_powered(const struct device *dev);
+ * pm_device_driver_init() if this is the case. Must be called at the beginning of a
+ * driver's deinit function.
+ *
+ * @note If CONFIG_PM_DEVICE=n, PM_DEVICE_ACTION_SUSPEND will be invoked
+ * unconditionally.
```
I find it weird how PM is being implemented when PM_DEVICE=n, i.e. in the header, all `pm_device_` calls are NOP when PM_DEVICE=n, this one and `pm_device_driver_init` however do "something" PM related. 
and fully in agreement about the above note as well, if PM_DEVICE=n do suspend using PM constructs?

Is this something that needs to be done outside of PM subsystem or is it just fallback and the primary use is when CONFIG_PM_DEVICE is enabled but where documentation need to be improved?

This probably is following the pre-existing `pm_device_driver_init` and trying to do `pm_device_driver_deinit` the same way, but all of this probably applies to `pm_device_driver_init` as well.

bjarki-andreasen replied:
> I find it weird how PM is being implemented when PM_DEVICE=n, i.e. in the header, all `pm_device_` calls are NOP when PM_DEVICE=n, this one and `pm_device_driver_init` however do "something" PM related. 
> and fully in agreement about the above note as well, if PM_DEVICE=n do suspend using PM constructs?
> 
> Is this something that needs to be done outside of PM subsystem or is it just fallback and the primary use is when CONFIG_PM_DEVICE is enabled but where documentation need to be improved?
> 
> This probably is following the pre-existing `pm_device_driver_init` and trying to do `pm_device_driver_deinit` the same way, but all of this probably applies to `pm_device_driver_init` as well.

PM_DEVICE=n does not mean users don't want their devices ACTIVE. We still need to resume devices, and the way to resume a device does not change regardless of when the application or kernel wants the device resumed. PM_DEVICE is all about runtime, it simply exposes the resume/suspend actions the driver will need to do anyway. pm_device_driver_init is a special case, where we get to directly call, thus reuse the drivers resume on init without exposing a hook for runtime use :)

decsny replied:
> > again strange for 2 reasons, first, it says just above that this function first ensures that it is already suspended, and 2nd, im still unclear why suspend action is ever being called if PM_DEVICE=n . My conceptualization of PM_DEVICE=n is that devices will ALWAYS be active in that case.  And actually this has serious implications I think to say now that devices can be suspended or have the suspend action invoked on them when PM_DEVICE=n . Because now what, do drivers have to #if CONFIG_PM_DEVICE some code in order to make the distinction between what they want to do with the suspend action depending on the configuration ? This seems relevant especially considering that it seems that pretty much every single pm-managed device driver is now also technically a domain driver and therefore any of these drivers could take actions which depower other devices upon getting suspended.
> 
> Device drivers do not behave differently whether PM_DEVICE=y or not, no ifdefery is needed.


We talked before on a different ticket about this and I explicitly made a table to clarify this and you added it to the RFC, which said that the devices never go out of active state when PM_DEVICE=n. That seems to be changing now. And I don't see how you can say that calling SUSPEND action on devices when PM_DEVICE=n makes any sense. The device driver is clearly going to potentially have suspending actions implemented and like you say, they shouldn't have any different behavior expected or ifdeffry based on PM_DEVICE=n. So this PR is seeming like it is just going to cause device PM to happen even when PM_DEVICE=n. It's not making sense to me and I'm not hearing a convincing explanation of why, and also not seeing an explanation of how even a driver is not going to be responding to suspend action with steps to actually suspend if PM_DEVICE=n. 

> Does it make sense for a device to be ACTIVE, if no driver is "loaded" for it? I say no :) With device_deinit() a device driver can be "unloaded", better suspend it using the driver before we unload it no?

With deinit we're talking about initialization state, not power state. Why are you making an equivalence between a device having a suspended power state and a device being deinitialized? Why coupling those two definitions exactly together? Instead, shouldn't it be symmetrical, where deinit just does the opposite of init, and suspend just does the opposite of resume? Does that mean that device init routine always does the same thing as device resume routine as well? I don't think so. So this is breaking symmetry, is unintuitive, and doesn't seem right to me, it seems to be doing something unexpected from multiple perspectives of different components and layers of the system.

nashif reviewed (changes requested):
lets get a few things clarified before we move forward with this.

bjarki-andreasen commented on include/zephyr/pm/device.h:
```diff
@@ -638,6 +638,23 @@ bool pm_device_is_powered(const struct device *dev);
+ * @param dev Device instance.
+ * @param action_cb Device PM control callback function.
+ * @retval 0 if success.
+ * @retval -EBUSY Device is not SUSPENDED nor OFF
```
Please read the section added to the docs regarding drivers without pm, it explains the rationale :)

bjarki-andreasen replied:
So, wording is confusing, by _Device Model without PM support_, its meant as _Device Model where internal PM operations are not exposed to the application_, so I believe _Device Model without Power Management Support_ is identical to _drivers with PM when PM_DEVICE=n_ 

The pm device docs introduction section has been updated to make this meaning explicit :)

bjarki-andreasen commented on include/zephyr/pm/device.h:
```diff
@@ -628,6 +628,9 @@ bool pm_device_is_powered(const struct device *dev);
  * never be powered off.
  *
+ * @note If CONFIG_PM_DEVICE=n, PM_DEVICE_ACTION_TURN_ON followed by
+ * PM_DEVICE_ACTION_RESUME will be invoked unconditionally.
```
How about

```
@note If CONFIG_PM_DEVICE=n, PM_DEVICE_ACTION_TURN_ON followed by
 * PM_DEVICE_ACTION_RESUME will be invoked on the device by this function.
```

decsny replied:
that's okay, I guess, although a pedantic linguistic/technicality nit might be that technically those actions are not "invoked on" the device, but maybe "presented to" or something similar. but either is fine for me.

bjarki-andreasen commented on include/zephyr/pm/device.h:
```diff
@@ -628,13 +628,33 @@ bool pm_device_is_powered(const struct device *dev);
 int pm_device_driver_init(const struct device *dev, pm_device_action_cb_t action_cb);
 
+/**
+ * @brief Deinit a device driver
```
How about

```
@brief Prepare PM device for device driver deinit
```

bjarki-andreasen replied:
changed to _Prepare PM device for device driver deinit_

decsny commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,47 @@ a particular device. It is important to emphasize that, although the
+  time of initialization, as these are the states within which device drivers
+  may be deinitialized.
+
+- ``SUSPENDED``: Device hardware is powered, but the device is not needed by the
+  system. The device should be put into its lowest internal power state,
+  commonly named "disabled" or "stopped". The device will be powered off if its
+  power domain is no longer needed by the system. Note that this state is
+  **NOT** a "low-power"/"partially operable" state, those are configured using
+  device driver specific APIs, and apply only while the device is ``ACTIVE``.
```
should you mention that the device's domain children would be powered off here?

decsny commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,47 @@ a particular device. It is important to emphasize that, although the
+  system. The device should be enabled in this state. Any device driver API may
+  be called in this state.
+
+- ``SUSPENDING``: Device hardware is powered, but the device has been enqueued to
+  be suspended, as it is no longer needed by the system. The device will be
+  dequeued in case the device becomes needed by the system, otherwise the device
+  will be suspended. No device driver API calls must occur in this state. Note
+  that this state is opaque to the device driver (no :c:enum:`pm_device_action`
+  is called as this state is entered)
```
this doesn't sound applicable to all systems, should an example be explained? why would there be a queue? this is something I actually never understood and neither has anyone I talked to about it so maybe an explanation of why this state exists, somewhere, would be good

bjarki-andreasen replied:
added the _why_ and `when` :)

ceolin replied:
This is very useful to avoid unnecessary power cycles for a device. Depending on who is getting a device it may not know if it will be used soon (imagine a subsystem) . If you are asking strictly about the state (and not the functionality) is pretty much because it is needed to implement the functionality described.

decsny commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +158,47 @@ a particular device. It is important to emphasize that, although the
 
-Each :c:enum:`pm_device_action` have a direct an unambiguous relationship with
-a :c:enum:`pm_device_state`.
+Each :c:enum:`pm_device_state` state is defined as follows:
```
I don't disagree with the contents of the documentation changes, and they are good changes, but why is it happening on this PR? what does it have to do with the pm_device_driver_deinit function being added?

bjarki-andreasen replied:
There is no wly attempt at anything in this PR :) Consider creating a new PR with only the docs, it will quickly stop making sense as pm_device_driver_deinit() is missing... or you could try rewriting the doc part without the deinit part, and then be stuck forever as atleast I would ask why we are missing this usecase :)

I will instead retitle the PR sine it has indeed changed focus a lot from when it was introduced, docs where added way later as requested by jordan

bjarki-andreasen replied:
changed title, and discussed PR at seconds arch meeting 20/6/2025

decsny commented on include/zephyr/pm/device.h:
```diff
@@ -628,13 +628,33 @@ bool pm_device_is_powered(const struct device *dev);
+ * @brief Deinit a device driver
+ *
+ * @details Ensures device is either SUSPENDED or OFF, and will revert the actions of
+ * pm_device_driver_init() if this is the case. Must be called at the beginning of a
+ * driver's deinit function.
```
It says must be called by driver, okay, but is it illegal to be called anywhere else ? if so, should that be clarified?

decsny reviewed (changes requested):
This is going to need more clarification about what is the difference or not between a device being suspended and deinitialized, right now this PR seems to make them equivalent, which I don't think should be the case. But really the device model maintainers need to weigh  in here about this, I don't know what was the use cases for why deinit was added in the first place, and I don't understand why is this PR seemingly equating deinit to a suspend power state in all cases.

bjarki-andreasen commented:
### Update

* Reworded introduction to clearly specify that PM_DEVICE is used to expose the device driver's PM operations outside of the driver
* Extended the SUSPENDED state doc entry to include the case where the device is itself a power domain
* Expanded the SUSPENDING state doc entry to explain why and when the state is used.
* Reworded header brief of pm_device_driver_deinit() from "Deinit a device driver" to "Prepare PM device for device driver deinit"
* Reworded header details of pm_device_driver_deinit() to not state that it "revert the actions of pm_device_driver_init"
* Reworded header notes for PM_DEVICE=n to replace "invoked unconditionally" "invoked on the device by this function".

diff: https://github.com/zephyrproject-rtos/zephyr/compare/ac7a10a1ef9302a945584418d95e4c2bba8473b2..39a33f541a9dac4eb8d2163dc7a60e92ec374280

ceolin commented:
@bjarki-andreasen I am particularly ok with the change, but I understand the friction here ...
There is some association between power state states and device init/de-init because in the end we are mapping a physical peripheral, but that makes drivers to implement some power management logic that they may not want to be triggered when PM is disabled.

It is not unusual to put `action_cb` under `CONFIG_PM_DEVICE` guard to save some resources if PM_DEVICE is disabled but this is not possible with this API, unless drivers also put the call to `pm_device_driver_deinit` under the same guard ... Same thing applies to `init` to be fair ... Can't we translate these operations to a NOP when PM is disabled ? Saying that because there may be a lot of additional code that is needed by the driver to properly implement power management actions that it may want to be out if `PM_DEVICE` is not used.

As I said, I don't have much problem with this change (specially the documentation are great) and I want to see more PM usage, but there are use cases where power is not exactly a problem but rom is limited, for example.

nashif reviewed (commented):
first batch of comments re driver PM states.  If the prupose of this PR is to define and agree on PM states as it was mentioned last week in the arch WG, then lets look at those first. One of the issues I see is how this PR is equating init/deinit with resume/suspend in different places, something we should not do IMO. This might be how someone defines their driver and I guess it should be possible, but should not be the main definition of how device driver PM states work.

More to follow re API addition for _deinit().

bjarki-andreasen commented:
> first batch of comments re driver PM states.  If the prupose of this PR is to define and agree on PM states as it was mentioned last week in the arch WG, then lets look at those first. One of the issues I see is how this PR is equating init/deinit with resume/suspend in different places, something we should not do IMO. This might be how someone defines their driver and I guess it should be possible, but should not be the main definition of how device driver PM states work.
> 
> More to follow re API addition for _deinit().

Thanks for the thorough feedback :) Please note that this PR is describing the status quo, and adding the missing counterpart to pm_device_driver_init(), so we have a baseline to work from, its not an attempt at refactoring PM :)

I agree with most of your feedback, yes our states differ from the D* states you describe, and yes, we need ways to define latency requirenments etc. for devices, and indeed SUSPENDING is not mappable to a "device lifecycle" state. This PR is not the place to address these though :)

Do you find anything in this PR inaccurate or misleading?

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
+  system. The device should be enabled in this state. Any device driver API may
+  be called in this state.
+
+- ``SUSPENDING``: Device hardware is powered, but the device has been scheduled
```
Agree, but not in this PR, this PR is about agreeing what this state means and why its there :) Which it seems we do

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
+  power domain is no longer needed by the system. If the device itself is a
+  power domain, devices on this power domain will be powered off in this state.
+
+  Note that this state is **NOT** a "low-power"/"partially operable" state,
```
As it is now, we have the state SUSPENDED and RESUMED. We may want to change that in the future, my proposal regarding device latency management was created because we have no way to define latency requirements for devices, but this PR is about agreeing on the status quo.

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
+  time of initialization, as these are the states within which device drivers
+  may be deinitialized.
+
+- ``SUSPENDED``: Device hardware is powered, but the device is not needed by the
```
As the doc states, it matches 
* It cannot perform any function 
* Device is mostly powered down.
as it is in its lowest power mode but the hardware has power. 

* Device still has auxiliary power (e.g., for wake-up signals).
is not a promise of this state, if it is in a state where it can monitor for signals and wake the system, it would be needed by the system, thus active, but in a low internal power state.

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
+- ``OFF``: Device hardware is not powered. This is the initial state from which
+  a device driver is initialized.
+
+  When a device driver is initialized, we do not know the state of the device.
```
I'm fairly confident this is primarily a semantics issue... Calling `pm_device_driver_init()` when `PM_DEVICE=n`, is not "using power management". Power management is explicitly `n`, the function relates to managing power states of a device, and thus is in `zephyr/pm/device.h`, along with the power states. This is the status quo, its perfectly functional and efficient, removing the need to duplicate the code required to move the device to an operating state, which is what happens when power management wants to enable a device.

Small correction, device goes from OFF to SUSPENDED to ACTIVE, and the device is not turning power on, the power domain which would be initialized before the device due to dependency ordinals would turn on the power for this device, the device is simply "reacting" to the power being on, necessarily after power was turned on (which would also be the case with power management). The procedure is identical to bring the device to an operating state, its just "fixed" if power management is not used.

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
+  a device driver is initialized.
+
+  When a device driver is initialized, we do not know the state of the device.
+  As a result, the :c:enum:`pm_device_action` ``TURN_ON`` action should be able
```
I think it does, if a user is debugging a driver, and wondering why TURN_ON is called after init, this would be a good bet as to where they would look I think, I'm okay with duplicating it somewhere else as well if you think there is a good place for it :)

decsny commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
+- ``SUSPENDED``: Device hardware is powered, but the device is not needed by the
+  system. The device should be put into its lowest internal power state,
+  commonly named "disabled" or "stopped". The device will be powered off if its
+  power domain is no longer needed by the system. If the device itself is a
+  power domain, devices on this power domain will be powered off in this state.
```
Is it not true that in this new scheme, all devices are treated like domains? And more accurate to say that if the domain has any children, they will be powered off? Is it more confusing to just say, that all pm-managed devices are domains, or that some are and some or not with no clear mechanism to describe that besides inference out of the lack of references to it?

bjarki-andreasen replied:
It is true that all power domains are devices, not that all devices are power domains :) The PM device context is the same for all devices, we check if a device (parent) has power domain children (`child { power-domains = <&parent>; };`) to determine if a specific device is a power domain, and if so, inform their children that the power domain was suspended (TURN_OFF). We don't treat busses like power domains for example just because they have child devices

decsny replied:
I think you are misunderstanding me here. I am not suggesting that something like a bus should be considered a domain due to having some sort of children of a different tree. I completely understand that the PM domain tree is an independent topological structure from DT, similar to the interrupts tree. In fact, what I am thinking of is even a device which has no type of children in any sense.  I think based on your architecture, which in my opinion is very clean and sound to allow for this, that any device which has declared PM resources, can be conceptually thought of as a power domain from the perspective of PM subsystem. It would simplify the wording of a lot of this and from what I can tell, is entirely true, these devices without power domain children are just domains which are the leaves of the power-domain tree, and from what I can tell, they are not actually supposed behaving any differently from "real" domain devices. They ought to turn themself off (in actual hardware) when they suspend and they can be notified of on/off from their parent when *it* turns off (even if only in theory in software state tracking).

What I am asking is what is the most simple way to think about this architecture that is still true so that it can be documented in a way that is most easy to understand. Because actually from what I understand of your proposed architecture, the behavior of them is the same as a device which is a real power domain, but that does not have children. There is not actually a difference in how the driver for a real power domain is written compared to any other device driver with PM resources declared now.  Even your default PM driver actually demonstrates this because it does nothing and is empty. I can imagine you could theoretically make as wide and deep of a tree of only these default empty power domain devices due to the fact that there is no special behavior from PM subsystem concept level for whether that device is a leaf or not in the domain tree.  So the same is true for any old device, from what I understand, even if they may have drivers that do a whole lot more than nothing, because as far as PM device are concerned, they are all acting the same as a domain would, actually.

bjarki-andreasen replied:
okay, we agree 100% on the power domain architecture, but the doc, as I have written it, makes it sound like power domains are a different "type" of device. I have reworded the section to describe "power domain devices" the exact same way as "normal" devices, particularly the wording `If a device has been specified as this device's power domain` should be very clear :)

decsny replied:
I guess that's fine what you have put now.  I would have preferred to be explicit and say directly that, any device which has PM resources declared becomes a power domain.  Which I understand is maybe a controversial way to conceptualize it.  But what you have changed it to is better enough than how it was before, IMO.

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
-a :c:enum:`pm_device_state`.
+Each :c:enum:`pm_device_state` state is defined as follows:
+
+- ``OFF``: Device hardware is not powered. This is the initial state from which
```
Its the initial state the device lifecycle starts in, always :) The device may be powered when we initialize, so the device hardware could indeed be SUSPENDED, but as the doc entry states, we can't know if the device lost power at any point before we init the driver, so we act as if it lost power, then gained power, then we init, and as such, we react as if the hardware did indeed loose power :)

nashif replied:
my point, this is a state, you define it as a state, how you got there involves a lot of logic and conditions that should not be part of the definition of what the state is. All of that can be left to the PM mode you are using. 

Same comment applies to all states below. 

But ok, lets see how this goes....

decsny commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,54 @@ a particular device. It is important to emphasize that, although the
+
+- ``SUSPENDING``: Device hardware is powered, but the device has been scheduled
+  to be suspended, as it is no longer needed by the system. This state is used
+  when delegating suspension of a device to the PM subsystem, optionally with
+  residency to avoid unnecessary suspend/resume cycles. The device will be
```
The wording is confusing to me here. Is it saying the residency is optional or this "delegated suspension" is optional and would always have residency.

Also, I am maybe lacking knowledge here, (perhaps that makes me qualified to review the documentation though), but I am not understanding how a device would get into the suspending state as opposed to going directly to suspend. And this residency that is mentioned, is it referring to the system mode residency or is there some other value associated for the device that is configured somehow?

decsny replied:
I think what you mean is that it is not mapped to a device *action* (as opposed to state). The actions are how the device driver makes sense of what it is supposed to do, and as you have pointed out, this state (suspending) does not seem to require any action on the part of a device.

decsny commented on include/zephyr/pm/device.h:
```diff
@@ -638,6 +638,23 @@ bool pm_device_is_powered(const struct device *dev);
+/**
+ * @brief Deinit a device driver
+ *
+ * @details Ensures device is either SUSPENDED or OFF, and will revert the actions of
```
Well, now I'm just confused, because on the first note it says that it ensures that the device is suspended, then the second note says it invoked suspend on it? That makes it read like this would never do anything, assuming the reason we are suspending is because it is not already? Or is the "ensure it is suspended" happening after this function was the cause of the suspend in the first place? It's the order here that is confusing perhaps, although I'm actually not sure I understand the intention of the API still just from reading this docstring

decsny reviewed (commented):
I have not had time to read all the new discussion on this thread but here is my check on some of my previous comments first

bjarki-andreasen commented:
@decsny Thanks for the review, indeed the docstring of `pm_device_driver_init()` and `pm_device_driver_deinit()` are a bit outdated now, the docs have been updated to (I think) clearly define their different behaviors when PM_DEVICE=y or n, I will add the same (but shortened) description to the docstrings :)

ceolin reviewed (dismissed):
I am not a super fan of having power states  being used with power management disabled but the way I see this is not being introduced here, `pm_device_driver_init` already does it and the way I see `pm_device_driver_deinit` is its counterpart. It seems that this end up saving code space because devices will end up re-using code for init/deinit and power management.

mmahadevan108 commented on doc/services/pm/device.rst:
```diff
@@ -158,8 +157,64 @@ a particular device. It is important to emphasize that, although the
+  may be deinitialized.
+
+  If a device has been specified as this device's power domain, and said device
+  becomes needed by the system, the power domain device will be transitioned
```
If the device becomes needed by the system, then shouldn't the device get transitioned to `ACTIVE` state as well?

bjarki-andreasen replied:
_the device_ here being the the device **on** the power domain right? The power domain device can become needed because some other device sharing the same power domain device became needed, _the device_ would only be moved all the way to ACTIVE if it became needed :)

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -4,11 +4,10 @@ Device Power Management
-setting :kconfig:option:`CONFIG_PM_DEVICE` to ``y``. When this option is
-selected, device drivers implementing power management will be able to take
-advantage of the device power management subsystem.
+Device Power Management (DEVICE PM) in Zephyr is a feature which makes
+device driver's PM operations available outside of the device driver,
+enabling power management of the device. This feature is enabled by
+selecting :kconfig:option:`CONFIG_PM_DEVICE`.
```
This change has been added to make it clear what the config PM_DEVICE means, and maybe more importantly, what it doesn't mean, as a common confusion brought up at arch meetings is that PM_DEVICE covers any and all PM operations a device driver must perform, as in, if PM_DEVICE=n, then a device driver has "no PM at all", which is simply not the case :)

But you are correct that this section seems like it describes Device Power Management as a whole now, which my change was not meant to to do, I am trying to describe specifically what the option PM_DEVICE does, to avoid the confusion mentioned before.

I like your description way better, I will add insert it, and move the description of the PM_DEVICE config :)

decsny commented on include/zephyr/pm/device.h:
```diff
@@ -620,13 +620,22 @@ int pm_device_power_domain_remove(const struct device *dev,
+ *   CONFIG_PM_DEVICE_RUNTIME=n) or (CONFIG_PM_DEVICE_RUNTIME=y and
+ *   !pm_device_runtime_is_enabled(dev)).
+ * - SUSPENDED if CONFIG_PM_DEVICE_RUNTIME=y and power domain is ACTIVE.
+ * - OFF if CONFIG_PM_DEVICE_RUNTIME=y and power domain is SUSPENDED.
```
Still could be interpreted wrong in the same way as I said

decsny reviewed (dismissed):
Changes of the overall pm device docs docs are basically fine improvements. The idea of having the pm_device_driver_deinit as the opposite of pm_device_driver_init which we already are using fairly ubiquitously is fine also.  But don't know why both of these things were done in one PR, I think it complicated the reviewing process by coupling them together.

bjarki-andreasen-nordic committed:
doc: releases: 4.3: mention pm_device_driver_deinit()

Mention introduction of pm_device_driver_deinit() in the 4.3 release
notes.

Signed-off-by: Bjarki Arge Andreasen <bjarki.andreasen@nordicsemi.no>

bjarki-andreasen commented:
### Update

Rebased and added release not entry for `pm_device_driver_deinit()` :)

nashif commented on doc/services/pm/device.rst:
```diff
@@ -202,58 +210,319 @@ power management implementation.
 
-        return 0;
-    }
+   struct dummy_driver_data {
```
That is a lot of code that sooner or later will be come stale , how about updating samples/subsys/pm/device_pm/, make it an annonated sample and instead reference the code from that sample here instead? This way we will know the code works and compiles and we have things in one place only.

teburd replied:
I've had this issue in other areas, I'd like to have compiled inline samples in the docs more or less, but the only way to do this today is make it a full fledged sample with independent docs and all of that... which makes no sense given its meant to be inline in the context of the docs.

I asked before about this before but we didn't come to any agreement on how best to do this.

Ideal scenario in my mind:
* inline doc samples go into a dedicated samples/doc_inline directory or something like it
* this directory contains samples meant to be *inlined* into docs and can therefore not be required to be documented as independent samples

nashif commented on include/zephyr/pm/device.h:
```diff
@@ -66,22 +66,76 @@ enum pm_device_flag {
 	 *
-	 * @note
-	 *     Device context may be lost.
+	 * @details The device should be put into its lowest internal power state,
```
I still think the details about transitions here do not belong into the state definition, but fine, we will see how this will work out and can always change it.

nashif replied:
yes, we need to rethink this as a whole and consider this option.

bjarki-andreasen commented on include/zephyr/pm/device.h:
```diff
@@ -622,13 +622,22 @@ int pm_device_power_domain_remove(const struct device *dev,
+ *
+ * The initial power state expected by the system is:
+ *
+ * - ACTIVE if CONFIG_PM_DEVICE=n or (CONFIG_PM_DEVICE=y and
```
I strongly expect using pseudo code to describe conditional logic is the opposite of "confusing and very difficult to understand" in a project tailored to software engineers...

bjarki-andreasen commented on doc/services/pm/device.rst:
```diff
@@ -202,58 +210,319 @@ power management implementation.
+           NULL
+   );
+
+Device Model without Power Management Support
```
"Device model driven" is not a phrase used anywhere in the project. Users would have no way of understanding what this section describes without reading the section first if we change the title.

Breaking the current title down, it has a good likelihood of being understood as the intended _"device driver model with PM_DEVICE=n"_

_"Device Model"_, as in device driver model, _"without Power Management"_, as in PM_DEVICE=n

bjarki-andreasen replied:
Changed to "Device Model without Device Power Management"

nashif replied:
PM_DEVICE kconfig help  "This option enables the device power management interface. .."

I wonder what this interface is exactly and if anything would actually work without this interface being enabled or used. Isnt 
```
static inline int pm_device_driver_deinit(const struct device *dev, pm_device_action_cb_t action_cb)
{
        return action_cb(dev, PM_DEVICE_ACTION_SUSPEND);
}
```
calling this interface  when CONFIG_PM_DEVICE=n , i.e.:
- PM_DEVICE_ACTION_SUSPEND is part of this interface
- pm_device_action_cb_t is also part of this interface 
- The implementation of this interface is guarded in most device drivers with `#ifdef CONFIG_PM_DEVICE`, which confirms the above.
- pm_device_driver_deinit itself is part of this interface

So you are using device power management in a different way. "without Device Power Management" is simply not true.

This is the main issue here and why every reviewer was "confused" and many of the comments in this PR are talking about confusion.

bjarki-andreasen commented:
### Update

Updated the titles
- _"Device Model with Power Management Support"_
- _"Device Model without Power Management Support"_

which primarily focus on the CONFIG_PM_DEVICE=y or n to specify _"Device Power Management Support"_ so they now read
- _"Device Model with Device Power Management Support"_
- _"Device Model without Device Power Management Support"_

bjarki-andreasen commented on doc/releases/release-notes-4.2.rst:
```diff
@@ -221,6 +221,7 @@ New APIs and options
     * :kconfig:option:`CONFIG_PM_DEVICE_RUNTIME_DEDICATED_WQ_PRIO`
     * :kconfig:option:`CONFIG_PM_DEVICE_RUNTIME_DEDICATED_WQ_INIT_PRIO`
     * :kconfig:option:`CONFIG_PM_DEVICE_RUNTIME_ASYNC`
+    * :c:func:`pm_device_driver_deinit`
```
updated :)

bjarki-andreasen commented:
Rebased and moved mention of introduction of `pm_device_driver_deinit` from release notes 4.2 to 4.3

mathieuchopstm reviewed (approved):
Let's merge this before the next conflict :slightly_smiling_face:
